### PR TITLE
CRI: Add kuberuntime container logs

### PR DESCRIPTION
**Pull request by Random-Liu**
Based on https://github.com/kubernetes/kubernetes/pull/34858.

The first 2 commits are from #34858. And the last 2 commits are new.

This PR added kuberuntime container logs support and add unit test for it.

I've tested all the functions manually, and I'll send another PR to write a node e2e test for container log.

**_Notice: current implementation doesn't support log rotation**_, which means that:
- It will not retrieve logs in rotated log file.
- If log rotation happens when following the log:
  - If the rotation is using create mode, we'll still follow the old file.
  - If the rotation is using copytruncate, we'll be reading at the original position and get nothing.

To solve these issues, kubelet needs to rotate the log itself, or at least kubelet should be able to control the the behavior of log rotator. These are doable but out of the scope of 1.5 and will be addressed in future release.

@yujuhong @feiskyer @yifan-gu 
/cc @kubernetes/sig-node

---

This change is [Reviewable](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/35348)

### CODE REVIEW

**Comment by Random-Liu** [comment]
@k8s-bot cri test this.

File: pkg/kubelet/kuberuntime/kuberuntime_logs.go [hunk 1]
```diff
@@ -0,0 +1,332 @@
...
+}
+
+// parseFunc is a function parsing one log line to the internal log type.
+// Notice that the caller must make sure logMessage is not nil.
+type parseFunc func([]byte, *logMessage) error
+
+var parseFuncs []parseFunc = []parseFunc{
+	parseCRILog,        // CRI log format parse function
+	parseDockerJSONLog, // Docker JSON log format parse function
+}
```
Comment (feiskyer) [inline pkg/kubelet/kuberuntime/kuberuntime_logs.go]: nit: Should doc those supported log formats (togather with delimiter and eol) in CRI interface.

File: pkg/kubelet/kuberuntime/kuberuntime_logs.go [hunk 2]
```diff
@@ -0,0 +1,332 @@
...
+	}
+
+	// Start parsing the logs.
+	r := bufio.NewReader(f)
+	// Do not create watcher here because it is not needed if `Follow` is false.
+	var w *fsnotify.Watcher
+	var parse parseFunc
+	msg := &logMessage{}
+	for {
+		l, err := r.ReadBytes(eol[0])
```
Comment (coffeepac) [inline pkg/kubelet/kuberuntime/kuberuntime_logs.go]: is there a way to specify a multi-character delimiter?  its probably not needed for stdout/stderr but for the future state of handling more arbitrary log files (stacktraces in an application log file jump to mind) it'd be nice to have multi-char event delimiters sorted out already.

if this is out of scope for now, please ignore.

**Comment by coffeepac** [comment]
One comment, otherwise looks good.

File: pkg/kubelet/kuberuntime/kuberuntime_logs.go (hunk 2 above)
Comment (Random-Liu) [inline pkg/kubelet/kuberuntime/kuberuntime_logs.go]: @coffeepac Currently, we only plan to support docker logging format and the CRI defined format.

In CRI defined format, runtime should decorate logs with timestamp and stream type, for example:
```
2016-10-06T00:17:09.669794202Z stdout log content 1
```
Currently, we assume `log content 1` is one line, and should be ended with `\n`. We may:
1) Assume all runtime should convert their logging into CRI format, which means that they may need to replace or escape their eol.
2) Support more logging format.

I think 1) is the current plan. @yujuhong

**Comment by Random-Liu** [comment]
I found that we already have a e2e test testing container log
[`should be able to retrieve and filter logs`](https://github.com/kubernetes/kubernetes/blob/master/test/e2e/kubectl.go#L775-L825). I think we could deprioritize the container logging node e2e test work.

File: pkg/kubelet/kuberuntime/kuberuntime_logs_test.go [hunk 3]
```diff
@@ -0,0 +1,191 @@
...
+		stdoutBuf := bytes.NewBuffer(nil)
+		stderrBuf := bytes.NewBuffer(nil)
+		err := writeLogs(msg, stdoutBuf, stderrBuf, &logOptions{since: test.since, timestamp: test.timestamp})
+		assert.NoError(t, err)
+		assert.Equal(t, test.expectStdout, stdoutBuf.String())
+		assert.Equal(t, test.expectStderr, stderrBuf.String())
+	}
+}
+
+func TestTail(t *testing.T) {
```
Comment (smarterclayton) [inline pkg/kubelet/kuberuntime/kuberuntime_logs_test.go]: Add a benchmark test too - the amount of memory allocated in this function and the behavior of logs will eventually be something we will tune.

File: pkg/kubelet/kuberuntime/kuberuntime_logs.go [hunk 4]
```diff
@@ -0,0 +1,332 @@
...
+		follow:    apiOpts.Follow,
+		timestamp: apiOpts.Timestamps,
+	}
+	if apiOpts.TailLines != nil {
+		opts.tail = *apiOpts.TailLines
+	}
+	if apiOpts.SinceSeconds != nil {
+		opts.since = now.Add(-time.Duration(*apiOpts.SinceSeconds) * time.Second)
+	}
+	if apiOpts.SinceTime != nil && apiOpts.SinceTime.After(opts.since) {
```
Comment (smarterclayton) [inline pkg/kubelet/kuberuntime/kuberuntime_logs.go]: Would recommend implementing maxBytes at this level.

File: pkg/kubelet/kuberuntime/kuberuntime_logs.go [hunk 5]
```diff
@@ -0,0 +1,332 @@
...
+var dockerJSONLog = &jsonlog.JSONLog{}
+
+// parseDockerJSONLog parses logs in Docker JSON log format. Docker JSON log format
+// example:
+//   {"log":"content 1","stream":"stdout","time":"2016-10-20T18:39:20.57606443Z"}
+//   {"log":"content 2","stream":"stderr","time":"2016-10-20T18:39:20.57606444Z"}
+func parseDockerJSONLog(log []byte, msg *logMessage) error {
+	dockerJSONLog.Reset()
+	l := dockerJSONLog
+	if err := json.Unmarshal(log, l); err != nil {
```
Comment (smarterclayton) [inline pkg/kubelet/kuberuntime/kuberuntime_logs.go]: This is hellishly expensive (I think).  Can you leave a TODO / open an issue to evaluate the performance of this later on?

Comment (Random-Liu) [inline pkg/kubelet/kuberuntime/kuberuntime_logs.go]: Docker logs also does json decoding, the only difference is that we are using `json.Unmarshal`, docker is using `json.Decoder` https://github.com/docker/docker/blob/b248de7e332b6e67b08a8981f68060e6ae629ccf/daemon/logger/jsonfilelog/read.go#L125.

Will there be a big performance difference between these 2?

File: pkg/kubelet/kuberuntime/kuberuntime_logs_test.go (hunk 3 above)
Comment (Random-Liu) [inline pkg/kubelet/kuberuntime/kuberuntime_logs_test.go]: Will do. BTW, I think the function will only use around `blockSize=1024b` memory, because we are always using the same buf - `buf := make([]byte, blockSize)`. It should be much better than docker's implementation, which read all logs into the memory https://github.com/docker/docker/blob/master/pkg/tailfile/tailfile.go#L20.

File: pkg/kubelet/kuberuntime/kuberuntime_logs.go (hunk 4 above)
Comment (Random-Liu) [inline pkg/kubelet/kuberuntime/kuberuntime_logs.go]: ACK. Will try to do this in `tail`

File: pkg/kubelet/kuberuntime/kuberuntime_logs.go (hunk 5 above)
Comment (smarterclayton) [inline pkg/kubelet/kuberuntime/kuberuntime_logs.go]: They're both equally bad.  This is fine for now but I'd like to replace this with a zero-allocation scanning approach later that avoids the expense of unmarshal.

File: pkg/kubelet/kuberuntime/kuberuntime_logs.go (hunk 4 above)
Comment (smarterclayton) [inline pkg/kubelet/kuberuntime/kuberuntime_logs.go]: maxBytes is particularly important for UIs to avoid blowing out browsers - doesn't have to be 100% accurate (original design allowed maxbytes to be slightly inaccurate), but is important to prevent DoS on kubelet as well.

File: pkg/kubelet/kuberuntime/kuberuntime_logs_test.go (hunk 3 above)
Comment (Random-Liu) [inline pkg/kubelet/kuberuntime/kuberuntime_logs_test.go]: @smarterclayton Ha, I see. I thought you were concerning about the performance of `tail`.

File: pkg/kubelet/kuberuntime/kuberuntime_logs.go (hunk 5 above)
Comment (Random-Liu) [inline pkg/kubelet/kuberuntime/kuberuntime_logs.go]: Hm, I see. I'll leave a TODO here.

File: pkg/kubelet/kuberuntime/kuberuntime_logs.go (hunk 1 above)
Comment (yujuhong) [inline pkg/kubelet/kuberuntime/kuberuntime_logs.go]: I think it's reasonable to return error for now.

File: pkg/kubelet/kuberuntime/kuberuntime_container.go [hunk 6]
```diff
@@ -675,14 +675,14 @@ func (m *kubeGenericRuntimeManager) AttachContainer(id kubecontainer.ContainerID
 
 // GetContainerLogs returns logs of a specific container.
 func (m *kubeGenericRuntimeManager) GetContainerLogs(pod *api.Pod, containerID kubecontainer.ContainerID, logOptions *api.PodLogOptions, stdout, stderr io.Writer) (err error) {
-	// Get logs directly from docker for in-process docker integration for
-	// now to unblock other tests.
-	// TODO: remove this hack after setting down on how to implement log
-	// retrieval/management.
-	if ds, ok := m.runtimeService.(dockershim.DockerLegacyService); ok {
-		return ds.GetContainerLogs(pod, containerID, logOptions, stdout, stderr)
+	status, err := m.runtimeService.ContainerStatus(containerID.ID)
```
Comment (yujuhong) [inline pkg/kubelet/kuberuntime/kuberuntime_container.go]: I don't think this is more reliable because you still rely on kubelet to pick a container ID to begin with. 
I am ok with this for now if we add a TODO comment for consolidating the code.

Comment (Random-Liu) [inline pkg/kubelet/kuberuntime/kuberuntime_container.go]: > I don't think this is more reliable because you still rely on kubelet to pick a container ID to begin with. 

Yeah, I just found that it's hard to get `RestartCount` correctly from `api.PodStatus` given the fragile weird logic, sometimes we need to use `RestartCount`, sometimes we need to use `RestartCount-1`. I think we should rely on the real log file in the future, if we separte the container and log lifecycle.

File: pkg/kubelet/kuberuntime/kuberuntime_logs.go (hunk 4 above)
Comment (Random-Liu) [inline pkg/kubelet/kuberuntime/kuberuntime_logs.go]: Done.

File: pkg/kubelet/kuberuntime/kuberuntime_logs_test.go (hunk 3 above)
Comment (Random-Liu) [inline pkg/kubelet/kuberuntime/kuberuntime_logs_test.go]: @smarterclayton The PR is already relatively big. I think the benchmark unit test won't block the PR itself that much. Are you ok with that I send another PR to add the benchmark test shortly?

**Comment by Random-Liu** [comment]
@k8s-bot cri test this.

File: pkg/kubelet/kuberuntime/kuberuntime_logs.go (hunk 1 above)
Comment (Random-Liu) [inline pkg/kubelet/kuberuntime/kuberuntime_logs.go]: ACK

**Comment by Random-Liu** [comment]
@k8s-bot cri test this.

**Comment by feiskyer** [comment]
lgtm

**Comment by coffeepac** [comment]
Review status: 0 of 7 files reviewed at latest revision, 6 unresolved discussions.

---

*[pkg/kubelet/kuberuntime/kuberuntime_logs.go, line 129 at r5](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/35348#-KUdl52NLZt_cnQs5n3D-r5-129:-KV7HaHEsthTToYfI6UO:b1v2fyv) ([raw file](https://github.com/kubernetes/kubernetes/blob/8a62d273ffeab3122c990890715b076d0e4158f6/pkg/kubelet/kuberuntime/kuberuntime_logs.go#L129)):*
<details><summary><i>Previously, yujuhong (Yu-Ju Hong) wrote…</i></summary>
> @coffeepac I think this is a valid request, and it's worth more discussion. We will not support this for the v1alpha1 CRI, but feel free to open an issue/proposal and cc sig-node.
</details>
@yujuhong  https://github.com/kubernetes/kubernetes/issues/35734

added

---


*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/35348)*

**Comment by Random-Liu** [comment]
*[pkg/kubelet/kuberuntime/kuberuntime_logs.go, line 129 at r5](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/35348#-KUdl52NLZt_cnQs5n3D-r5-129:-KV7LcsKUOpUen_xJMnB:b-ucfknh) ([raw file](https://github.com/kubernetes/kubernetes/blob/8a62d273ffeab3122c990890715b076d0e4158f6/pkg/kubelet/kuberuntime/kuberuntime_logs.go#L129)):*
<details><summary><i>Previously, coffeepac (choused) wrote…</i></summary>
> @yujuhong  https://github.com/kubernetes/kubernetes/issues/35734
> 
> added
</details>
Thanks!

---


*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/35348)*

**Comment by yujuhong** [comment]
LGTM overall. 

@smarterclayton are you ok with a separate PR for the benchmarking?

**Comment by smarterclayton** [comment]
Yes, we can put it on the CRI beta list

On Oct 28, 2016, at 6:16 PM, Yu-Ju Hong <notifications@github.com> wrote:

LGTM overall.

@smarterclayton <https://github.com/smarterclayton> are you ok with a
separate PR for the benchmarking?

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub
<https://github.com/kubernetes/kubernetes/pull/35348#issuecomment-257042007>,
or mute the thread

**Comment by Random-Liu** [comment]
@k8s-bot unit test this https://github.com/kubernetes/kubernetes/issues/35898

**Comment by Random-Liu** [comment]
@k8s-bot gci gke e2e test this

**Comment by Random-Liu** [comment]
@k8s-bot unit test this #35898

**Comment by Random-Liu** [comment]
@k8s-bot unit test this #35898
